### PR TITLE
Using Babel for TypeScript instead of tsc

### PR DESCRIPTION
**Pull request by kroeder**
Issue: https://github.com/storybooks/storybook/issues/5030

## What I did

Using babel as typescript compiler instead of tsc. This allows us to rely on only babel and also gives us the possibility to use a prop-type generator on compile time

### CODE REVIEW

File: .babelrc.js
Comment (igor-dv) [inline .babelrc.js]: I assume we need this only for angular, so let's have an override for angular with this (see the `overrides` down here)

Comment (kroeder) [reply .babelrc.js]: Vue.js with TS benefits from this as well https://vuejs.org/v2/guide/typescript.html#Class-Style-Vue-Components

Also, using decorators is a general TypeScript feature - does one babel plugin costs so much that it is worth to only use them for certain projects?

Comment (igor-dv) [reply .babelrc.js]: Nah, just legacy: true makes me uncomfortable. But nvm then

Comment (kroeder) [reply .babelrc.js]: Check this link https://babeljs.io/docs/en/next/babel-plugin-proposal-decorators.html#legacy
I don't know what stage 1 or stage 2 is

I also don't know how decorators work in stage 2
If you think we should apply these only to vue and angular then lets do this.

But as said it does not work.. It compiles but brings up runtime errors in Storybook/Angular :(

**Comment by Hypnosphi** [comment]
Do we still need a separate `compile-ts.js` after this change?

File: .babelrc.js
Comment (ndelangen) [reply .babelrc.js]: What are the runtime errors?

Comment (kroeder) [reply .babelrc.js]: Removed this line - for now, if we have a project that needs decorators, we use `tsc` instead of babel

**Comment by kroeder** [comment]
@Hypnosphi I still need to do a couple of things like renaming files

After having some trobules with angular and decorators in general (so no blame on angular except the architectural choice of using decorators 😛 ) we ended up using babel for js + ts except for angular that now uses `tsc` 

I need to thing about a name now
- `compile-js` does babel 
- `compile-ts` does tsc stuff (not only for angular but also for generating declaration files  `d.ts`)

Maybe we just rename them `compile-babel` and `compile-tsc` ?

**Comment by kroeder** [comment]
For the record: I had to migrate @storybook/angular in order to be able to compile it with `tsc` instead of babel because lots of stuff in this package is written in `js`

**Comment by maraisr** [comment]
AFIK Babel's TypeScript does not give you the _benefits_ of TypeScript, ie compile-time type checking. It just pulls out the types so babylon can understand it.

**Comment by kroeder** [comment]
@maraisr good point! In a guide I found it actually is advertised as feature because it makes the dev watch-mode faster. But the same guide said that you can add `"check-types": "tsc"` to your package.json scripts

We should think about how and when we do want type checks. 
- As pre-commit like our linting rules? 
- As something a dev has to do manually
- As a `concurrent` process, same as `yarn dev` runs `dev:js` and `dev:ts` at the moment
- Something that just happens in the pipeline (I do not prefer this, it's way too late and can be troublesome when doing bigger changes)

I personally could imagine to put it in a separate process or as pre-commit because these errors possibly cause bugs in run-time

Doing it manually won't work IMO. Devs might oversee this

**Comment by benoitdion** [comment]
>As a concurrent process, same as yarn dev runs dev:js and dev:ts at the moment
That's a pretty nice experience and should be more than enough. vscode continues to give the same level of support even when using babel.  https://github.com/Realytics/fork-ts-checker-webpack-plugin is one way to achieve this.

**Comment by ndelangen** [comment]
Working on this

**Comment by ndelangen** [comment]
@benoitdion Would you be able to help with this one?

**Comment by benoitdion** [comment]
I can try. Is one package giving you grief in particular? any reason you upgraded to typescript 3.4 RC?

**Comment by benoitdion** [comment]
I spent a bit of time on this and unfortunately didn't get to a resolution. Some observation you maybe  already be aware of:
- seems the main culprits are angular related.
- we probably shouldn't be on typescript 3.4 yet
- `storyshot-core` is mostly all javascript. The error when running the tests seems to originate from there. Possibly a red herring.
- as far as I can tell, we're including the right presets for angular. 
- lots of ts errors if I run tsc over the entire project. `yarn dev:tsc` is more conservative but also shows a few errors.
- we should turn on `isolatedModules` everywhere, it helps catch babel incompatibilities in ts code. (see caveats section in https://devblogs.microsoft.com/typescript/typescript-and-babel-7/)

**Comment by kroeder** [comment]
Except netlify, everything seems to be good now. Added `do not merge` because there's an error during `yarn dev` that needs to be solved first

File: package.json
Comment (shilman) [inline package.json]: @kroeder Why does `yarn dev` still spawn `yarn dev:tsc`? I thought the major selling point of this PR was that we wouldn't have to run both?

**Review by shilman — CHANGES_REQUESTED** [review]
I saw the comment about angular needing `tsc` so perhaps that's why we're running both in `yarn dev`? If that's the case, is there a way to minimize the overhead of running `tsc`?

**Comment by shilman** [comment]
Did a timing of `yarn bootstrap --reset --core` on this branch and on `next`:

This PR:
```
real	8m9.690s
user	9m47.683s
sys	4m39.407s
```

`next` branch:
```
real	7m48.649s
user	6m27.471s
sys	4m40.299s
```

Are there any opportunities to speed this up?

**Comment by kroeder** [comment]
Can you please confirm my changes by using `time yarn bootstrap --core` without `--reset`? Installing node_modules is affected by network time and can vary on each run 

A few tests based on my last commits
```
Tests are based on: time yarn bootstrap --core


Test-run with no changes
------------------------

real    2m17.811s
user    0m0.060s
sys     0m0.170s


Run only remove-dist + babel
----------------------------

real    0m53.101s
user    0m0.030s
sys     0m0.121s

After trying a lot I saw no potential in decreasing the babel-build time


Run only remove-dist + tscfy 
----------------------------

- No changes
real    1m5.509s
user    0m0.076s
sys     0m0.045s

- Adding "skipDefaultLibCheck": true
real    0m58.443s
user    0m0.075s
sys     0m0.107s

- adding "skipLibCheck": true besides "skipDefaultLibCheck": true
real    0m36.928s
user    0m0.075s
sys     0m0.106s


Final result after all optimizations
------------------------------------

real    1m30.265s
user    0m0.045s
sys     0m0.090s
```

**Comment by kroeder** [comment]
`next` is still faster 
```
real    1m13.757s
user    0m0.030s
sys     0m0.230s
```

I can try a couple more things and check the differences between `next` and my branch but I start to question if this PR is a benefit for Storybook at all if we.
- Cannot get rid of `tsc` because we need typings
- Cannot get it faster due to some unknown reasons

Initially I wanted to move it to babel to have 
- Only one build system
- Being able to generate react prop types by using typescript interfaces using https://github.com/milesj/babel-plugin-typescript-to-proptypes (I'm aware that this package has not much attention and I need to test if it works properly)

File: examples/angular-cli/package.json
Comment (benoitdion) [inline examples/angular-cli/package.json]: 3.4 is now out 🎉

File: addons/storyshots/storyshots-core/src/frameworks/angular/loader.js
Comment (benoitdion) [inline addons/storyshots/storyshots-core/src/frameworks/angular/loader.js]: looks like you might be able to remove some of this now

File: examples/cra-ts-kitchen-sink/tsconfig.json
Comment (benoitdion) [inline examples/cra-ts-kitchen-sink/tsconfig.json]: using `isolatedModules` in all our tsconfigs could improve tsc performance

File: lib/ui/src/keyboard/keyCodes.ts
Comment (benoitdion) [inline lib/ui/src/keyboard/keyCodes.ts]: can this be removed?

File: scripts/compile-babel.js
Comment (benoitdion) [inline scripts/compile-babel.js]: do you want to keep this set to true after the PR gets merged?

File: scripts/compile-babel.js
Comment (benoitdion) [inline scripts/compile-babel.js]: Very good find! Hopefully this gets easier in the future

File: scripts/compile-tsc.js
Comment (benoitdion) [inline scripts/compile-tsc.js]: still needed?

File: scripts/compile-tsc.js
Comment (benoitdion) [inline scripts/compile-tsc.js]: maybe this should reference back the comment in `compile-babel`

File: package.json
Comment (benoitdion) [reply package.json]: shouldn't this be `dev:babel` + `dev:check=types` and only `dev:tsc` for angular?

Comment (kroeder) [reply package.json]: Dev:tsc creates definition files and compiles angular, check types just checks types with --noEmit

File: examples/cra-ts-kitchen-sink/tsconfig.json
Comment (leoyli) [reply examples/cra-ts-kitchen-sink/tsconfig.json]: should we also add `declaration: false` here?! I'm not sure what is the best way to make sure this one won't emit any typed files.  If we want an emission, then this have to be set to false.  See this issue on [TS repo here](https://github.com/Microsoft/TypeScript/issues/29490).

**Comment by leoyli** [comment]
Q: Why we want to inject react prop types? For people not using TS or for react-docgen?

**Comment by shilman** [comment]
@kroeder Seems like the performance is comparable to `next`.

This PR:
```
real	1m52.390s
user	4m9.852s
sys	0m26.077s
```

`next`:
real	2m3.644s
user	3m45.516s
sys	0m56.512s

File: scripts/compile-babel.js
Comment (kroeder) [reply scripts/compile-babel.js]: good catch

File: addons/storyshots/storyshots-core/src/frameworks/angular/loader.js
Comment (kroeder) [reply addons/storyshots/storyshots-core/src/frameworks/angular/loader.js]: @ndelangen you changed that, can you take a look at this?

File: examples/cra-ts-kitchen-sink/tsconfig.json
Comment (kroeder) [reply examples/cra-ts-kitchen-sink/tsconfig.json]: `isolatedModules: true`
```real    0m18.465s
user    0m0.045s
sys     0m0.045s```

`isolatedModules: false`
real    0m18.345s
user    0m0.030s
sys     0m0.091s```

Why should it be faster? What does this option even do? I tried to find more accurate documentation than "exports every file as separate module"

Also keep in mind that this is an example app and depends on the decisions a project owner makes, not storybook.

@ndelangen you added `isolatedModules: true` - Were you just playing around with the options or was this on purpose?

Comment (benoitdion) [reply examples/cra-ts-kitchen-sink/tsconfig.json]: it could/should be faster because tsc doesn't need cross-file information to emit. Doesn't look like it makes any difference for us though. 

I would still recommend adding it in because it helps enforce "Babel's single file emit architecture". See the caveats section at https://devblogs.microsoft.com/typescript/typescript-and-babel-7/

File: scripts/compile-tsc.js
Comment (kroeder) [reply scripts/compile-tsc.js]: There are log outputs in the compile scripts that are marked red without

File: examples/angular-cli/package.json
Comment (kroeder) [reply examples/angular-cli/package.json]: Updated to 3.4.1

File: lib/ui/src/keyboard/keyCodes.ts
Comment (kroeder) [reply lib/ui/src/keyboard/keyCodes.ts]: Just talked to @ndelangen 
It came in by merging next into this branch - I'll leave it as is because it might be needed by the code owner

File: scripts/compile-tsc.js
Comment (kroeder) [reply scripts/compile-tsc.js]: Thanks! Added a comment in my latest commit

File: package.json
Comment (kroeder) [reply package.json]: @shilman  In a green dream world I tried to just run `tsc` after each babel-compile.js run which means there are only 50 instead of 100 watch-processes (which is insane 💥 ) but I noticed that I cannot hook on `babel --watch` so `tsc --watch` needs to be started as well for generating `d.ts` files

I thought about to use a custom file watcher that just starts `babel` and `tsc` without `--watch` but I was worried about the performance because babel + tsc would need a cold-start without `--watch`. `tsc` has a cache since 3.4 but I'm not sure how babel handles cold starts

Ultimately, I thought this can be step 2 in improving the performance and I wanted to finish this first step since it already took me a lot of time to do this change

**Comment by kroeder** [comment]
> Q: Why we want to inject react prop types? For people not using TS or for react-docgen?

@leoyli I'm not a react user but when I first started the whole typescript migration I had to write types twice:

- Interfaces for TS users
- PropTypes for non-TS users

These can get out of sync so my intention was to get rid of the manually written prop-types

**Comment by leoyli** [comment]
@kroeder, I can feel your pain...  Thank your these great works and your explains.

I think it is really questionable on where we need to use prop-types. According to [React official doc](https://reactjs.org/docs/static-type-checking.html), it seems that the prop-types is really redundant if we decide to migrate the whole project into TS.  But that is out of the scope here... 😓

File: examples/cra-ts-kitchen-sink/tsconfig.json
Comment (kroeder) [reply examples/cra-ts-kitchen-sink/tsconfig.json]: I just found out: ` TS5053: Option 'declaration' cannot be specified with option 'isolatedModules'.` 🐙

Comment (leoyli) [reply examples/cra-ts-kitchen-sink/tsconfig.json]: That is why I suggest to add `declaration: false` together with `isolatedModules`, maybe with a comment?  In my own project, I have `tsc --isolatedModules --noEmit --watch` for type checking, and `tsc --declaration --removeComments --emitDeclarationOnly` for type emission. The rest of them goes through babel.

File: addons/storyshots/storyshots-core/src/frameworks/angular/loader.js
Comment (ndelangen) [reply addons/storyshots/storyshots-core/src/frameworks/angular/loader.js]: Should remove if storyshots run successfully without it

File: examples/angular-cli/.storybook/webpack.config.ts
Comment (ndelangen) [inline examples/angular-cli/.storybook/webpack.config.ts]: mixing es modules & commonjs here, is that recommended?

File: examples/cra-ts-kitchen-sink/tsconfig.json
Comment (ndelangen) [reply examples/cra-ts-kitchen-sink/tsconfig.json]: I'm with @leoyli on this one, sounds like that's what we want?

File: package.json
Comment (ndelangen) [reply package.json]: Implementing a watcher ourselves would be a bad idea.

We should make sure that both tsc & babel watch as few files as possible

File: scripts/compile-tsc.js
Comment (ndelangen) [reply scripts/compile-tsc.js]: next PR: replace tslint with eslint everywhere, then it can be removed

**Review by ndelangen — APPROVED** [review]

**Comment by ndelangen** [comment]
> if we decide to migrate the whole project into TS

That's not out of scope, it's our roadmap.

But never the less, users of storybook and addon-creators are not all migrating.
So by adding proptypes to our storybook components they'll be able to catch problems and help us detect issues at runtime too.

File: tsconfig.json
Comment (gaetanmaisse) [inline tsconfig.json]: Do you set incremental build to `false` because it is not working with babel for now?

Comment (kroeder) [reply tsconfig.json]: I did that because it failed in a certain scenario.. I think calling `yarn bootstrap --core` twice
It should not use a cache here because it removes dist and compiles everything from scratch

I just forgot to add that option for `yarn dev` in watch mode
Maybe we can do a tsconfig.prod.json and a tsconfig.json to add transparency in the tsc build process

**Comment by kroeder** [comment]
If it is okay for everyone I'd like to do a feature-freeze for this PR and collect them in an issue for some sort of "Babel refactoring - Phase II" and only fix bugs or change requests that should be done before merging (if we decide to merge 🙂 ) - Is that okay?

**Currently on the list:**
- Make `yarn dev` more efficient and less CPU consuming
- Make `tsc` build transparent (by e.g. adding context based tsconfig.json, tsconfig.prod.json etc.)
- Try adding `isolatedModules: true` when `declarations` can be `false`
- <Add your proposal here>

File: examples/cra-ts-kitchen-sink/tsconfig.json
Comment (kroeder) [reply examples/cra-ts-kitchen-sink/tsconfig.json]: Hm question:
- You start `yarn dev` and `yarn dev:tsc` needs to create declaration files for other internal packages that rely on that package: How can I do `declaration: false` here
- You start `yarn bootstrap --core` and internally packages again rely on the `d.ts` files of other packages: How can I do `declaration: false`

Is there something I don't get or is this a misunderstanding? 🙂

Comment (kroeder) [reply examples/cra-ts-kitchen-sink/tsconfig.json]: I mean, yes - it won't fail but at least your IDE will mark those imports red because they are missing - Isn't it confusing if you get red lines in your IDE?

File: package.json
Comment (kroeder) [reply package.json]: Fun fact: babel uses `gulp watch`
What transpiles babel? 😮 🙂

File: examples/angular-cli/.storybook/webpack.config.ts
Comment (kroeder) [reply examples/angular-cli/.storybook/webpack.config.ts]: Is it not recommended? A common linting rule in `ts` is `no-use-require` or something like that. I can revert that if you want, it works though. I never expected any trouble using imports for path like that

**Review by shilman — APPROVED** [review]

**Comment by libetl** [comment]
this one will have a conflict with my (very small) PR : https://github.com/storybooks/storybook/pull/6415/files 

Do you think you can make the same change as in mine ? @kroeder 
Thank Kai.

**Comment by ndelangen** [comment]
Let's merge @kroeder ?

**Comment by ndelangen** [comment]
Congrats @kroeder 

Epic work

File: examples/angular-cli/.storybook/webpack.config.ts
Comment (mohsinulhaq) [reply examples/angular-cli/.storybook/webpack.config.ts]: https://github.com/webpack/webpack/issues/4039#issuecomment-273804003
what about this?

Comment (kroeder) [reply examples/angular-cli/.storybook/webpack.config.ts]: This post is 2 years old and maybe babel / typescript already converts my import to es5 (can't check right now)

Comment (mohsinulhaq) [reply examples/angular-cli/.storybook/webpack.config.ts]: I came across this issue yesterday itself where I was mixing the two. This led to me getting some imports inside `default` and some not. Not mixing them up helped me fix it. Although I am not quite sure about your usage.

**Review by Hypnosphi — COMMENTED** [review]
Hi @kroeder! Do you remember why did you have to revert?
https://github.com/storybookjs/storybook/pull/5109/commits/6e6df7ccfb9a19743372ab60c754aa0f2b14d650

**Comment by kroeder** [comment]
@hypnosphi Mh no I'm not sure but can try to figure this out tomorrow. Does it cause any trouble?

**Comment by Hypnosphi** [comment]
Not really, just curious

**Pull request merged**
